torch/_dynamo/variables/builtin.py

```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
             tx.output.side_effects.is_attribute_mutation(obj)
             and name_var.is_python_constant()
         ):
+            if isinstance(obj, variables.TensorVariable):
+                from .builder import wrap_fx_proxy
+
+                if name_var.as_python_constant() == "data":
+                    # Remove the old reference in tracked fakes - if we don't do this
+                    # new .data value size and shape differences will cause
+                    # tracked fakes to produce incorrect guards. This is sound because the TensorVariable
+                    # coming out of set_() below will be a new one, and get
+                    # installed in tracked fakes.
+                    to_remove = []
+                    for tf in tx.output.tracked_fakes:
+                        if tf.source == obj.source:
+                            to_remove.append(tf)
+                    for tf in to_remove:
+                        tx.output.tracked_fakes.remove(tf)
+
+                    # Step 1 - disable grad
+                    version = obj.as_proxy().node.meta["example_value"]._version
+                    with dynamo_disable_grad(tx), torch.no_grad():
+                        # Step 2 - call `set_`
+                        out = wrap_fx_proxy(
+                            tx,
+                            tx.output.create_proxy(
+                                "call_function",
+                                torch.Tensor.set_,
+                                *proxy_args_kwargs([obj, val], {}),
+                            ),
+                        )
+                    # Step 3 - drop the version counter - this is a hack required to get
+                    # .data setting to play correctly with the autograd engine.
+                    tx.output.create_proxy(
+                        "call_function",
+                        torch._C._autograd._unsafe_set_version_counter,
+                        (out.as_proxy(), 0),
+                        {},
+                    )
+                    # Step 4 - replace all reference to the current object with the new one
+                    tx.replace_all(obj, out)
+                    return val.add_options(self, out, name_var)
```

Comment:
I think don't return here. Setattr is supposed to return None, and we also need to track the setattr as a side effect.

Reply:
So we need the code outside of this block to run

Reply:
Actually, do we? Hmm. I've got a feeling it's the responsibility of AOT autograd to persist the `set_` mutation

Reply:
Returning None doesn't matter here, it gets handled correctly later iirc, but we can make that change.

Object tracking can be updated a little, sure, not sure it matters tho.

Reply:
Right, I think this part is fine actually. The problem is with AOT Autograd. I think tests are failing due to this: https://github.com/pytorch/pytorch/issues/110842

Reply:
Let me see if I can try to handle it.